Upstatement/ups-editorial-wp-plugin
# v2

This PR refactors and updates the `ups-editorial` plugin (previously, `ups-blocks`) to account for recent core updates and best practices. It also removes dependencies on any WordPress theme (most notably the [WP Starter Kit](https://github.com/Upstatement/wordpress-starter-kit)) so as to fully function like a standalone plugin.

The changes introduced in this plugin correlate to many changes introduced to the WP Starter Kit in that [repo's v2](https://github.com/Upstatement/wordpress-starter-kit/pull/203).

This plugin offers the following features:

- Extends a number of blocks, primarily to limit the block's customization options available to editors. These blocks include the Cover, File, Gallery, Image Layout, Related Articles, Table, and Video blocks.
- Adds two editor plugins that create new panels in the Gutenberg sidebar:
    - The Article Topper Panel allows an editor to control article topper metadata.
    - The Bylines Panel improves an editor's ability to manage a post's authors by offering an interface to add multiple ordered authors, driven by a custom `ups_author` taxonomy registered by the plugin.
- Exposes template functions that can be used to access data registered by the plugin.
- Allows plugin functionality to be configured (including the ability to enable and disable specific features) via a config file that can be added to the root of the active theme directory. This file should have the filename `ups-editorial.php` and simply return an array of relevant configuration values (see the README in this branch for documentation on the specific configuration values).

## Changes from the previous version

- Refactors the plugin's structure and namespacing, including a more organized approach to the features defined in PHP vs. those defined in JS.

- Removes the custom `ups/image` block. As WordPress continued to improve their editor experience and update their core blocks, the customizations intended by the `ups/image` block became antiquated as core image block functionality became more configurable.

    > I could see a world where the updates intended by this custom block are useful, but I'd like for any functionality introduced by a future potential `ups/image` block to be very intentional and backed by established requirements.

- Simplifies how JavaScript and stylesheets are built and loaded. Previously, there were a number of disparate JavaScript files and scss partials floating around, hooked into a custom webpack config file. This update removes the custom webpack config by routing all JavaScript functionality through a singular `src/index.js` file, which is how the `wp-scripts` package expects the code to be organized.

## How to test

1. Download the code in the `v2` branch of this repo and add it to an existing WordPress installation. (The plugin should work for any WordPress installation running the latest version of the Gutenberg editor, but the [`v2` branch of Upstatement's WP Starter Kit](https://github.com/Upstatement/wordpress-starter-kit/pull/203) is a good blank slate to test against.)

2. In the plugin's directory, run `npm install` to install dev dependencies, then run `npm run build` to compile the production assets. You'll also need to run `composer install` to generate the autoloader (this will also install some dev dependencies like linters). Note that the intention is for this plugin to be pulled into a WordPress environment via composer, so release versions of this package will be shipped with compiled assets and any relevant dependencies all ready to go. This configuration is happening in a separate PR - this one is primarily about the specific functionality of the plugin.

3. Activate the plugin.

4. If your theme has a `ups-editorial.php` config file, delete it. This will ensure that all plugin functionality is active.

5. Confirm that an `Authors` taxonomy exists for your posts. Confirm that the bylines functionality works as expected (it should continue to work as it did before):

    - You should be able to add author taxonomy terms via the taxonomy's regular admin page (`/wp-admin/edit-tags.php?taxonomy=ups_authors`)
    - You should be able to search for and select individual author terms in the `Authors` panel when editing a post. You should also be able to reorder and delete the author terms added via this panel.
    - You should be able to update the authors via the quick edit menu on the main posts admin screen.

6. Confirm that the `Article Topper` panel in the editor works as expected. You should be able to add topper content and switch between the different article topper variations.

7. Confirm that the Cover, File, Gallery, Image Layout, Related Articles, Table, and Video blocks all work as expected and have the expected customizations (they should continue to work in the same way as before). Note that the Image Layout and Related Articles blocks are _not_ core blocks and need to be defined in your theme - these are defined in the WP Starter Kit. Future work on this plugin may include a better way to handle these non-core blocks.

8. Confirm that the template methods work in your theme. See the `Theme API` section of the README for this branch for documentation on which template methods are available.
